rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,158 @@
+discovering a cluster/installing kube-agent/creating a service account, we will
+create cluster role bindings for those roles, accordingly linking them to the groups "default-cluster-admin", "default-view" and
+"default-edit". It will give users an opportunity to always have standard set of Kubernetes permissions which they can use together
+with Teleport's fine-grained RBAC definitions for Kubernetes. This will make enrolling and starting using Kubernetes clusters
+into Teleport a much simpler process, since user will be able to setup RBAC completely in Teleport if they want. For example,
+role giving read-only access to staging namespace pods might look like this:
+
+```yaml
+kind: role
+metadata:
+  name: staging-kube-access
+version: v7
+spec:
+  allow:
+    kubernetes_labels:
+      'region': '*'
+      'platform': 'aws'
+    kubernetes_resources:
+      - kind: pod
+        namespace: "staging"
+        name: "*"
+    kubernetes_groups:
+    - default-view
+    kubernetes_users:
+    - developer
+  deny: {}
+```
+
+Users will not need to create a separate role binding in Kubernetes for it to work - once the cluster is enrolled in Teleport, the default-view is already there.
+We will add an option to the teleport-kube-agent chart and to our
+kubeconfig generation script to not create those cluster role bindings, but it will be enabled by default. 
+
+### Automatically provisioning RBAC resources to Kubernetes clusters.
+
+We will add a new type of resources, called "KubeProvision", where user will be able to specify Kubernetes RBAC resources they want to
+be provisioned onto their Kubernetes clusters.
+clusters. A single Teleport resource of that type can define multiple Kubernetes RBAC resources. There's a limit on size of a resource Teleport
+can save on the backend, therefore, if some users would have unusually large amount of Kubernetes RBAC resources they want to provision, they would
+need to split it into two KubeProvision resources.
+
+```protobuf
+import "teleport/header/v1/metadata.proto";
+
+// KubeProvision represents a Kubernetes resources that can be provisioned on the Kubernetes clusters.
+// This includes roles/role bindings and cluster roles/cluster role bindings.
+// For rationale behind this type, see the RFD 174.
+message KubeProvision {
+  // The kind of resource represented.
+  string kind = 1;
+  // Not populated for this resource type.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+  // The specific properties of kube provision.
+  KubeProvisionSpec spec = 5;
+}
+
+// KubeProvisionSpec is the spec for the kube provision message.
+message KubeProvisionSpec {
+  // resources_data is base64 encoded YAML definitions of the Kubernetes resources.
+  string resources_data = 3;
```

Comment:
can we make this less general so people that need to create them have a nice experience.

Reply:
Yeah, I'm looking into trying to make it more user-friendly.